### PR TITLE
cleanup memory protection Kconfigs and unnecessary dependencies

### PR DESCRIPTION
**Pull request by tyrantlizard77**
User mode has evolved over the years, particularly the scope and intended usage of memory domains. This PR addresses a dependency issue in the ARM code, and tidies up some Kconfigs.

- The ARM code no longer directly depends on the `k_mem_partition` outside of `CONFIG_USERSPACE`. A new `struct z_arm_mpu_partition` is defined to specifically capture the contents of MPU region registers. The MPU code now uses this structure internally.

Note: De-coupling this is necessary to allow for flexibility in maintenance of the userspace feature, without worrying about pervasively breaking arch/ code that isn't related to userspace. A follow-up PR will take a look at `k_mem_partition_attr_t`; I want to decouple these defines from USERSPACE or memory domains and allow these to be used with `z_mem_map()`, for example.

- The core kernel `struct k_mem_partition` definition is now behind `CONFIG_USERSPACE` as intended.

- A new CONFIG_MPU Kconfig has been added. It simply indicates that an MPU is active and gets selected by the architecture.

- CONFIG_SRAM_REGION_PERMISSIONS and CONFIG_MEMORY_PROTECTION have been merged into CONFIG_SRAM_REGION_PERMISSIONS and is now selectable by either MMU or MPU systems. It indicates that global permissions have been set up for regions within the kernel image.

Note: Our existing MPU implementations unconditionally set up these regions, so they are selected by ARC_MPU or ARM_MPU. There are some edge use-cases on MMU systems *not* to enable this, so it is a default-y option for those systems.

### CODE REVIEW

**Comment by andrewboie** [comment]
I tested this on our QEMU ARM targets and also frdm_k64f to test the NXP MPU changes.

**Comment by carlocaione** [comment]
I tested this on top of #29551 enabling `CONFIG_SRAM_REGION_PERMISSIONS` and it works correctly.

**Review by carlocaione — APPROVED** [review]

**Comment by andrewboie** [comment]
@ioannisg let me know what you think of this new arrangement. I removed CONFIG_MEMORY_PROTECTION, but want to make sure I'm not missing any semantics in the kernel. Since that protection test depended on it, and that test is evaluating static region permissions, does that capture the intention here or do you have any additional use-cases for this config?

for example, maybe a convenience config which is selected if we have either CONFIG_MMU or CONFIG_MPU, but I'm not sure what would use this since I don't think we have any APIs conditional on this.

**Review by nashif — APPROVED** [review]

**Comment by carlescufi** [comment]
@andrewboie please rebase
@ioannisg please review

File: arch/Kconfig [hunk 1]
```diff
@@ -642,13 +633,14 @@ config KERNEL_VM_SIZE
 
 endif   # MMU
 
-config MEMORY_PROTECTION
```
Comment (andrewboie) [inline arch/Kconfig:645]: note: not currently seeing use-cases for CONFIG_MEMORY_PROTECTION, but if we wanted to keep it, could be something as simple as:

```
config MEMORY_PROTECTION
	bool
	default y
	depends on MMU || MPU
```

**Pull request merged**